File: npc_sim/perception/perception_system.py
# Copyright 2025-2026 Sadık Abdusselam Albayrak
# Licensed under the Apache License, Version 2.0
"""Per-NPC perception orchestrator."""

from __future__ import annotations
from npc_sim.core.sim_vector3 import SimVector3
from npc_sim.events.stimulus import Stimulus, StimulusType
from npc_sim.npc.vitals import NPCVitals
from npc_sim.npc.psychology import NPCPsychology
from npc_sim.perception.perceived_object import PerceivedObject
from npc_sim.perception.sensor_range import SensorRange
from npc_sim.perception.perception_filter import PerceptionFilter


class PerceptionSystem:
    """
    Per-NPC perception orchestrator. Each tick it filters stimuli through
    sensor range and salience evaluation, upserts percepts, and prunes expired ones.
    """

    def __init__(self, sensor: SensorRange = None, pfilter: PerceptionFilter = None):
        self.sensor = sensor or SensorRange()
        self.filter = pfilter or PerceptionFilter()
        self._percepts: list[PerceivedObject] = []
        self.percept_timeout: float = 30.0

    @property
    def active_percepts(self) -> list[PerceivedObject]:
        return self._percepts

    def tick(self, stimuli: list[Stimulus], npc_position: SimVector3,
             npc_forward: SimVector3, vitals: NPCVitals,
             psychology: NPCPsychology, current_time: float) -> list[PerceivedObject]:
        # Mark all existing percepts as not currently visible
        for p in self._percepts:
            p.mark_not_visible()

        changed: list[PerceivedObject] = []

        for stimulus in stimuli:
            # Spatial pre-filter
            if not self.sensor.can_sense(stimulus, npc_position, npc_forward):
                continue

            # Salience evaluation
            salience = self.filter.evaluate(stimulus, vitals, psychology)
            if salience <= 0.0:
                continue

            # Derive threat level
            threat = 0.0
            if stimulus.tag.lower() == "threat":
                threat = stimulus.intensity * (0.5 + psychology.neuroticism * 0.5)

            # Upsert percept
            existing = self._find_percept(stimulus.source_id)
            if existing is not None:
                existing.refresh(stimulus.source_position, current_time, threat, salience)
            else:
                obj_type = self._infer_object_type(stimulus)
                existing = PerceivedObject(
                    stimulus.source_id, obj_type, stimulus.source_position,
                    current_time, threat, salience, stimulus.tag)
                self._percepts.append(existing)

            changed.append(existing)

        # Prune expired
        self._percepts = [p for p in self._percepts
                          if not p.is_expired(current_time, self.percept_timeout)]

        return changed

    def get_threats(self) -> list[PerceivedObject]:
        result = [p for p in self._percepts
                  if p.tag.lower() == "threat" and p.threat_level > 0]
        result.sort(key=lambda p: p.threat_level, reverse=True)
        return result

    def get_allies(self) -> list[PerceivedObject]:
        return [p for p in self._percepts if p.tag.lower() == "ally"]

    def get_nearest_food(self, npc_position: SimVector3) -> PerceivedObject | None:
        nearest = None
        nearest_sqr = float("inf")
        for p in self._percepts:
            if p.tag.lower() != "food":
                continue
            d = SimVector3.sqr_distance(npc_position, p.last_known_position)
            if d < nearest_sqr:
                nearest_sqr = d
                nearest = p
        return nearest

    def _find_percept(self, source_id: str) -> PerceivedObject | None:
        for p in self._percepts:
            if p.object_id == source_id:
                return p
        return None

    @staticmethod
    def _infer_object_type(s: Stimulus) -> str:
        if s.type == StimulusType.VISUAL:
            return "Hazard" if s.tag == "Threat" else "Entity"
        elif s.type == StimulusType.SOCIAL:
            return "NPC"
        elif s.type == StimulusType.AUDIO:
            return "Noise"
        elif s.type == StimulusType.OLFACTORY:
            return "Scent"
        return "Unknown"

    def __repr__(self) -> str:
        return f"[PerceptionSystem] {len(self._percepts)} active percept(s)"

File: Runtime/Perception/PerceptionSystem.cs
using System;
using System.Collections.Generic;

namespace ForgeProject.Sim
{
    /// <summary>
    /// Per-NPC perception orchestrator.
    /// Each tick it:
    ///   1. Pre-filters stimuli by sensor range (cheap distance check).
    ///   2. Evaluates salience via <see cref="PerceptionFilter"/> (psychology-aware).
    ///   3. Upserts / creates <see cref="PerceivedObject"/> entries.
    ///   4. Prunes expired percepts.
    ///   5. Returns changed percepts so the caller can route them to NPC.WitnessEvent().
    ///
    /// Held per-NPC. Stateful between ticks (retains percepts until they expire).
    /// </summary>
    public sealed class PerceptionSystem
    {
        // ─── Configuration ────────────────────────────────────────────────────────

        /// <summary>How many sim-seconds a percept survives without refresh.</summary>
        public float PerceptTimeout { get; set; } = 30f;

        // ─── Components ───────────────────────────────────────────────────────────

        private readonly SensorRange      _sensor;
        private readonly PerceptionFilter _filter;

        // ─── State ────────────────────────────────────────────────────────────────

        private readonly List<PerceivedObject> _percepts = new();

        /// <summary>All currently active perceived objects.</summary>
        public IReadOnlyList<PerceivedObject> ActivePercepts => _percepts;

        // ─── Constructor ──────────────────────────────────────────────────────────

        public PerceptionSystem(SensorRange sensor = null, PerceptionFilter filter = null)
        {
            _sensor = sensor ?? new SensorRange();
            _filter = filter ?? new PerceptionFilter();
        }

        // ─── Tick ─────────────────────────────────────────────────────────────────

        /// <summary>
        /// Processes a batch of stimuli for a single tick.
        /// </summary>
        /// <param name="stimuli">All stimuli available this tick (pre-dispatched to this NPC).</param>
        /// <param name="npcPosition">Current world position of the NPC.</param>
        /// <param name="npcForward">Facing direction (Zero = omnidirectional).</param>
        /// <param name="vitals">NPC vitals used by the filter.</param>
        /// <param name="psychology">NPC psychology used by the filter.</param>
        /// <param name="currentTime">Current simulation timestamp.</param>
        /// <returns>List of percepts that were newly created or updated this tick.</returns>
        public List<PerceivedObject> Tick(
            IReadOnlyList<Stimulus> stimuli,
            SimVector3   npcPosition,
            SimVector3   npcForward,
            NPCVitals    vitals,
            NPCPsychology psychology,
            float        currentTime)
        {
            // Mark all existing percepts as "not currently visible" before this tick
            foreach (var p in _percepts) p.MarkNotVisible();

            var changed = new List<PerceivedObject>();

            foreach (var stimulus in stimuli)
            {
                // Step 1: Spatial pre-filter (cheap)
                if (!_sensor.CanSense(stimulus, npcPosition, npcForward)) continue;

                // Step 2: Salience evaluation (psychology-based)
                float salience = _filter.Evaluate(stimulus, vitals, psychology);
                if (salience <= 0f) continue;

                // Derive threat level from intensity + tag
                float threat = stimulus.Tag.Equals("Threat", StringComparison.OrdinalIgnoreCase)
                    ? stimulus.Intensity * (0.5f + psychology.Neuroticism * 0.5f)
                    : 0f;

                // Step 3: Upsert percept
                var existing = FindPercept(stimulus.SourceId);
                if (existing != null)
                {
                    existing.Refresh(stimulus.SourcePosition, currentTime, threat, salience);
                }
                else
                {
                    string objectType = InferObjectType(stimulus);
                    existing = new PerceivedObject(
                        stimulus.SourceId, objectType, stimulus.SourcePosition,
                        currentTime, threat, salience, stimulus.Tag);
                    _percepts.Add(existing);
                }

                changed.Add(existing);
            }

            // Step 4: Prune expired percepts
            _percepts.RemoveAll(p => p.IsExpired(currentTime, PerceptTimeout));

            return changed;
        }

        // ─── Convenience queries ──────────────────────────────────────────────────

        /// <summary>Returns all active percepts tagged as threats, sorted by threat level descending.</summary>
        public List<PerceivedObject> GetThreats()
        {
            var result = _percepts.FindAll(p =>
                p.Tag.Equals("Threat", StringComparison.OrdinalIgnoreCase) && p.ThreatLevel > 0f);
            result.Sort((a, b) => b.ThreatLevel.CompareTo(a.ThreatLevel));
            return result;
        }

        /// <summary>Returns all active percepts tagged as allies.</summary>
        public List<PerceivedObject> GetAllies() =>
            _percepts.FindAll(p => p.Tag.Equals("Ally", StringComparison.OrdinalIgnoreCase));

        /// <summary>Returns the closest food percept, or null.</summary>
        public PerceivedObject GetNearestFood(SimVector3 npcPosition)
        {
            PerceivedObject nearest = null;
            float nearestSqr = float.MaxValue;

            foreach (var p in _percepts)
            {
                if (!p.Tag.Equals("Food", StringComparison.OrdinalIgnoreCase)) continue;
                float d = SimVector3.SqrDistance(npcPosition, p.LastKnownPosition);
                if (d < nearestSqr) { nearestSqr = d; nearest = p; }
            }
            return nearest;
        }

        // ─── Helpers ──────────────────────────────────────────────────────────────

        private PerceivedObject FindPercept(string sourceId)
        {
            foreach (var p in _percepts)
                if (p.ObjectId == sourceId) return p;
            return null;
        }

        private static string InferObjectType(Stimulus s) => s.Type switch
        {
            StimulusType.Visual    => s.Tag == "Threat" ? "Hazard" : "Entity",
            StimulusType.Social    => "NPC",
            StimulusType.Audio     => "Noise",
            StimulusType.Olfactory => "Scent",
            _                      => "Unknown"
        };

        public override string ToString() =>
            $"[PerceptionSystem] {_percepts.Count} active percept(s)";
    }
}
